Add Path Tracer

philippkoelmel opened the pull request:
Hey You!

From the docs:
> Sometimes it might be helpful to see the path the player took. That can help to analyse the situation. There are several options:
> 
> - Record your movement: If `pathtracer_record_player` is set to `1` then the path and movement keys are recorded. You can switch this on and off as you like. You can bind this to some keys or mouse buttons
> - Show your movement: `pathtracer_show_player` to `1` draws the previously recorded information into the current scene.
> - Show the path of a ghost or demo: `pathtracer_show_demo` on `1` will draw the path of the demo/ghost. It can draw the movement keys if the demo contains that information
> - Fadeout: `pathtracer_fadeout_ghost` and `pathtracer_fadeout_demo` if set to `1` will fadeout the path around the current position. This helps to keep an overview if the path gets a bit convoluted. Per default the fadeout is disabled for ghosts, enabled for demos. The time the path will be drawn is defined by `pathtracer_fadeout_seconds` in seconds (defaults to `3`).

I have iterated a few times since recording this video: https://www.youtube.com/watch?v=_ALKW1LBp0Y

Actually it helps me now to better understand how I can improve my bunnyhops. But turns out watching demos with the path tracer was also fun, so I added support for not only the ghost, but also demos. For that I had to move some ghost code from private to public. 

The code is not well tested yet. I have not tried marathons or other multi-map demos. I did not test multiplayer demos. So still some stuff to do before this is ready. But I thought it might already be helpful to open the PR since the code base is getting bigger and bigger.

philippkoelmel commented:
Tested with a coop demo on a custom map and a marathon.

https://www.youtube.com/watch?v=SwooZjeV0x0

![image](https://github.com/user-attachments/assets/c617ddfa-eb21-47a5-af3c-9398b0f3b538)

j0zzz commented:
Hey Philipp!

Thank you for your work, I have compiled your PR and shared with a smaller group of runners to try it out and tell us what they think about it. You know, at first we always like to analyze and discuss if any change would give some unfair advantage for speedrunners. Your feature for me seems like a nice visual enhancement, which I kindly welcome. I believe it can be helpful for newcomers trying to learn especially the bunnyhopping basics and map routes.

If you ask me, I would surely like to add it into JoeQuake.

philippkoelmel commented:
Hi Joe! Thanks, that makes me really happy to hear! I hope there are no negative side effects and it all works out for the runner, looking forward to their feedback!

matthewearl commented:
I like the per-frame stripes on the path, it makes it easy to see in one picture various things.  Some issues I spotted after a quick play:

- There is a break in the path when 10fps is used. ![image](https://github.com/user-attachments/assets/c8e1c2c1-7dcb-4fbc-9a0f-1b7e16ad2a59)
- The path isn't drawn when sv_player is unset.  This means that if you run `playdemo` immediately after starting the game without having started a map, then the path isn't drawn.

It'd also be neat to be able to compare the demo with the ghost when you're playing a demo.  Plus I'm not sure the benefit of having the live player path shown, but maybe I'm missing a use case?  

Had a quick look at the code and perhaps it might be nicer just to directly include ghost_private.h in your pathtracer.c rather than expose the internals.  Ideally the ghost code wouldn't be used for non-ghost purposes but I understand this would require more refactoring than you probably want to do.

> This adds additional load time

Is the load time noticeable?  The level load time normally dominates the ghost load time although with dzip demos there can be a pause.  We might want to lazily load the path data then.

On the whole this is very cool for demo analysis, and should also be useful for learning routes (by following the ghost's path).  If you can think of a way to load the paths from the demos without invoking the internal ghost code that would be even better.

philippkoelmel commented:
> There is a break in the path when 10fps is used.

If the path (vector length) between two frames is longer than 50 units then the code skips drawing the line. The idea was to avoid the lines across the whole map when teleporting. But looks like with 10fps that can happen with normal movement. Will check soon!

> The path isn't drawn when sv_player is unset. This means that if you run playdemo immediately after starting the game without having started a map, then the path isn't drawn.

Thx, checking soon!

> It'd also be neat to be able to compare the demo with the ghost when you're playing a demo. 

Makes sense! Will add this.

> Plus I'm not sure the benefit of having the live player path shown, but maybe I'm missing a use case?

That helped me to review what I actually did during bunny hops. I compare what the ghost is doing to what I have been doing. Then I can see if I press the right buttons on the right frames, move the mouse in the right direction, etc. Then I can try again, record another section of the path and compare it again. My brain just needs this visualisation of what I actually did to fully understand what's going on.

> include ghost_private.h in your pathtracer.c rather than expose the internals

I use the ghost code in cl_demo.c and in the pathtracer.c in cl_demo.c (https://github.com/j0zzz/JoeQuake/pull/189/files#diff-e045a11c62e8b3c4676a74119f348297eb0fe37d0d4c4d93be5df44936b63170R573). It is a short cut to extract the player movement from the demo. On refactoring, see below.

> Is the load time noticeable? 

When I load the big marathons demos (all levels) I noticed a few freeze frames, but I have not tested loading this demo wo the pathtracer. Maybe it is normal? 
I can change the code, it would read through the whole demo file, extract player movement information and then reset the read head to the beginning of the demo and let the existing code takes over. I am not sure if that is meaningfully faster though. But I could make the ghost code private again.
Alternatively I could only track what the existing code has extracted from the demo. This means there would not be a path that shows the future movement.
I can even make a toggle to alternate between these two behaviors.

I hope I got all your points!

matthewearl commented:
> When I load the big marathons demos (all levels) I noticed a few freeze frames, but I have not tested loading this demo wo the pathtracer. Maybe it is normal?

When I play one of the big demos (eg `all_5716.dz`)  it stalls while extracting the dzip but the actual ghost loading code is quick..

> That helped me to review what I actually did during bunny hops.

I think have the having the demo path and the ghost path visible when playing a demo might be the better interface for this, and it'd keep the code simpler too.   So you'd do your practice run, then review the demo to see how it compares.  What do you think?

> I use the ghost code in cl_demo.c and in the pathtracer.c in cl_demo.c

I see.  I'm happy with the way you have it now, although I might make a PR in the future to separate the path loading part out of the ghost module so that it can be used in both places without having to invoke the ghost code in a module that isn't ghost related.

> Alternatively I could only track what the existing code has extracted from the demo. This means there would not be a path that shows the future movement.

Don't do that, being able to see the path ahead is super cool!

By the way, have you tested the demo path viewing with a plain .dem file, not in a .dz?   For some reason it is not working for me (no path appears) but I have not dug into it.

philippkoelmel commented:
> By the way, have you tested the demo path viewing with a plain .dem file, not in a .dz? For some reason it is not working for me (no path appears) but I have not dug into it.

That should work. I tested with extracted and self recorded demos. Can you share the demo you tested with?

> I think have the having the demo path and the ghost path visible when playing a demo might be the better interface for this, and it'd keep the code simpler too. So you'd do your practice run, then review the demo to see how it compares. What do you think?

Checking soon!

philippkoelmel commented:
> It'd also be neat to be able to compare the demo with the ghost when you're playing a demo.

You can now play the demo, see the path. Then add a ghost and see both paths. The ghost becomes transparent when it comes close to the player. If both, demo and ghost, are very similar then you won't see the ghost.

> There is a break in the path when 10fps is used.

Fixed via `pathtracer_skip_line_threshold` and updated docs accordingly.

> The path isn't drawn when sv_player is unset. This means that if you run playdemo immediately after starting the game without having started a map, then the path isn't drawn.

Still open.

> I think have the having the demo path and the ghost path visible when playing a demo might be the better interface for this, and it'd keep the code simpler too. So you'd do your practice run, then review the demo to see how it compares. What do you think?

For me the live path tracing is the key feature :-) It helps me (a super noob) to have a very tight feedback loop of bunnyhopping. Immediately after a few bunnies (not even making it to the end of the map) I can compare my behavior to the ghost. 

But, in case the code complexity is your primary concern, I was able to simplify the code by just re-using the ghost structs and removing the path tracer custom structs. The draw code is now the same for demo, ghost or live path tracing. The player live recording code is also much simpler.

(I accidentally committed some project files, removed them and force pushed)

philippkoelmel commented:
> By the way, have you tested the demo path viewing with a plain .dem file, not in a .dz? For some reason it is not working for me (no path appears) but I have not dug into it.

I might have an idea! Checking...

philippkoelmel commented:
> By the way, have you tested the demo path viewing with a plain .dem file, not in a .dz? For some reason it is not working for me (no path appears) but I have not dug into it.

Fixed. I never used the `playdemo` command. Instead I always went through the demo menu. The command handles the path differently thus I had to use the COM_FOpenFile instead fopen. That fixed the path problem and is now in line with how demos are usually loaded.

> The path isn't drawn when sv_player is unset. This means that if you run playdemo immediately after starting the game without having started a map, then the path isn't drawn.

Fixed.

I am going to clean up some more files in the next commits.

philippkoelmel commented:
> Update: Not fixed, still at it.

Got it now. I had a workaround with the demo file name. I removed this and replaced it with the (already existing) file stream. For that I had to change the ghost code to not close the file if the caller wishes.

I tried to make the changes as small as possible. I guess sooner or later I should squash the commits.

philippkoelmel commented:
PS: Not sure if I introduced that bug. If you play a demo with movekeys and then play a demo without movekeys then the last state of the movekeys from the previous demo are drawn.

matthewearl commented:
Hey, these updates look great, thanks.  Plain dems are now working for me and the `sv_player` fix is working well too.   I'd be tempted to make the `pathtracer_skip_line_threshold` default even higher since in situations like the e3m5 water dunk at 10fps you can jump more than 80 units a frame, but it's not a big deal:

![image](https://github.com/user-attachments/assets/23eed6cc-8ed0-4f05-a0d1-cf7a4ebf5916)

On the file handle closing change, I think there would be an issue with what I call "chained" demos.  These are marathon demos that are stored as separate demos, that link to each other with a stuffcmd `playdemo <next demo>` at the end.  Older marathons (older than a few years) plus all qdq releases tend to use them. 

I reckon there will be an issue when you play one of these chained demos.  In this case, `Ghost_ReadDemo` would be called from `StartPlayingOpenedDemo` and then we end up leaking a load of file descriptors:  It will loop through the whole chain of demos, without closing any of them.    What I suggest as a fix is:
- Get rid of the `shall_close_file` flag.
- Expose `Ghost_ReadDemoNoChain` and call this directly from `StartPlayingOpenedDemo` instead of `Ghost_ReadDemo`.
- Move the `fclose()` call out of `Ghost_ReadDemoNoChain` and into `Ghost_ReadDemo`.

Again, this change is really cool.  It makes it really clear where the player takes a different route from the ghost:

![image](https://github.com/user-attachments/assets/88a328f7-be90-4f25-a500-4d550c0be5cb)

philippkoelmel commented:
> you can jump more than 80 units a frame

Incredible! :-) I will increase to 100 units!

> Get rid of the shall_close_file flag. Expose Ghost_ReadDemoNoChain and call this directly from StartPlayingOpenedDemo instead of Ghost_ReadDemo. Move the fclose() call out of Ghost_ReadDemoNoChain and into Ghost_ReadDemo.

I see. I have a lot to learn here :-) Great idea I will do that.

>  It makes it really clear where the player takes a different route from the ghost

Interesting!

Did you also see the movekeys attached to the path? Is that also helpful? I can add a toggle if that is too confusing.

matthewearl commented:
> Did you also see the movekeys attached to the path? Is that also helpful? I can add a toggle if that is too confusing.

Yes, looks like a neat feature for figuring out involved moves.   A toggle sounds like a good idea though!

philippkoelmel commented on trunk/cl_demo.c:
```diff
@@ -540,13 +547,23 @@ void StartPlayingOpenedDemo (void)
+
+	demo_offset = ftell (cls.demofile);
+	demo_info = Q_calloc (1, sizeof(*demo_info));
+	Ghost_ReadDemoNoChain (cls.demofile, demo_info, next_demo_path);
```
works for me, tested dzip, dem, playdemo and main menu.

philippkoelmel commented on trunk/ghost/ghostparse.c:
```diff
@@ -413,6 +423,12 @@ Ghost_ReadDemo (FILE *demo_file, ghost_info_t **ghost_info_p)
+            fclose(demo_file);
+            demo_file = NULL;
+        }
+
```
moved over here. I think the intent of Ghost_ReadDemoNoChain is to exit the function with a closed file.

philippkoelmel commented:
> Yes, looks like a neat feature for figuring out involved moves. A toggle sounds like a good idea though!

added.

philippkoelmel commented:
> Get rid of the shall_close_file flag. Expose Ghost_ReadDemoNoChain and call this directly from StartPlayingOpenedDemo instead of Ghost_ReadDemo. Move the fclose() call out of Ghost_ReadDemoNoChain and into Ghost_ReadDemo.

Worked well. Added some comments to the code.

matthewearl reviewed (commented):
I left a few minor comments, but it looks good to me.  Thanks for taking the time to implement this.  What do you reckon @j0zzz ?

philippkoelmel commented on trunk/pathtracer/pathtracer.c:
```diff
@@ -0,0 +1,389 @@
+	}
+}
+
+static void PathTracer_Draw_Level (ghost_level_t* level, boolean fadeout_enable, float fadeout_seconds, float skip_line_threshold, boolean show_movekeys)
```
done

philippkoelmel commented on trunk/pathtracer/pathtracer.c:
```diff
@@ -0,0 +1,389 @@
+	glPolygonMode(GL_FRONT_AND_BACK, GL_LINE);
+	glDisable(GL_TEXTURE_2D);
+	glDisable(GL_CULL_FACE);
+	boolean gl_line_smooth_was_enabled = glIsEnabled(GL_LINE_SMOOTH);
```
-> qboolean, done.

j0zzz commented:
@philippkoelmel very nice job, thank you for your efforts! @matthewearl thanks for the tips and insights!
I have tried it out once again, it looks very cool. I like that it's highly configurable, great to have many options. This is usually requested by the community members, to have several options to fine-tune their environments.

If you think it is in a feature complete state, I will merge it into the master branch.

philippkoelmel commented:
Thx Joe! I am ready to go!
